### PR TITLE
live update of acquisition

### PR DESCRIPTION
**Pull request by Kleijwegt**
provides live image
works for: CL, SEM, ARMDstream, streak camera
orange next pixels work
however has a blue NON-transparant background
Still 9 points to discuss with Eric [See TODO K.K. ERIC]

### CODE REVIEW

File: src/odemis/acq/stream/_sync.py [hunk 1]
```diff
@@ -774,6 +779,149 @@ def _stopLeeches(self):
...
+        rgbim[numpy.invert(self._acq_mask), 1] = 121
+        rgbim[numpy.invert(self._acq_mask), 2] = 192
+
+        if hasattr(self, "_npixels2scan"):
+            size_preview = self._npixels2scan
+        else:
+            size_preview = 1
+
+        # Orange progress pixels
+        idx = numpy.where(self._acq_mask == False)
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: I don't think this is very reliable. In particular, I think in case of "fuzzy", this is not going to work. An alternative which might work could to have a _scanning_area which is basically a left/top/rigth/bottom (either as 4 int, or a numpy slice), which would directly fit on the SEM data. It would be computed/updated by the runAcquisition function.

File: src/odemis/acq/stream/_sync.py [hunk 2]
```diff
@@ -774,6 +779,149 @@ def _stopLeeches(self):
...
+        rgbim[idx[0][0:size_preview],idx[1][0:size_preview],0] = 204
+        rgbim[idx[0][0:size_preview], idx[1][0:size_preview], 1] = 130
+        rgbim[idx[0][0:size_preview], idx[1][0:size_preview], 2] = 0
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: You can write directly the last dimension, so:
`rgbim[idx[0][0:size_preview],idx[1][0:size_preview]] = (204, 130, 0)`
That magical number should be moved to a constant like `SCANNING_AREA_COLOUR`.

File: src/odemis/acq/stream/_sync.py [hunk 3]
```diff
@@ -1062,18 +1206,32 @@ def _runAcquisitionEbeam(self, future):
...
+                                                     args = (weakref.ref(self), self._live_update_period),
+                                                     name = "Live update of the acquisition")
+            self._liveImageThread.daemon = True
+            self._liveImageThread.start()
+
+
+            for pol_idx, pol_pos in enumerate(pos_polarizations):
                 if pol_pos is not None:
-                    logging.debug("Acquiring with the polarization position %s", pol_pos)
+                    logging.debug("Acquire with the following polarization position: %s" % pol_pos)
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Please "revert", as for the other logging lines.

File: src/odemis/acq/stream/_sync.py [hunk 4]
```diff
@@ -1062,18 +1206,32 @@ def _runAcquisitionEbeam(self, future):
...
                     f = self._analyzer.moveAbs({"pol": pol_pos})
                     f.result()
                     time_move_pol_left -= time_move_pol_once
 
-                # iterate over pixel positions for scanning
+                # iterate over pixel positions for scanning.
                 for px_idx in numpy.ndindex(*rep[::-1]):  # last dim (X) iterates first
                     trans = tuple(spot_pos[px_idx])  # spot position
 
+                    self._current_scan_area = (px_idx[1] * tile_size[0],
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Let's define it in the `__init__`, for documentation, as a tuple `(0,0,0,0)`.

File: src/odemis/acq/stream/_sync.py [hunk 5]
```diff
@@ -1118,7 +1275,19 @@ def _runAcquisitionEbeam(self, future):
...
+                    for i, das in enumerate(self._acq_data):
+                        self._assembleLiveData(i, das[-1], px_idx, rep, pol_idx)
+
+                    # Activate _updateImage thread
+                    self._shouldUpdateImage()
+
+                    # Live update the setting stream with the new data
+                    self._sccd._onNewData(self._ccd_df, self._acq_data[self._ccd_idx][-1])
+
+                    logging.debug("Done acquiring image number %s out of %s." % (n, tot_num))
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Typically, in logging, it's considered nicer to read to have each variable passed as a separate parameter (as it used to be for this line). So no need to change to using `%`.

File: src/odemis/acq/stream/_sync.py [hunk 6]
```diff
@@ -1130,8 +1299,11 @@ def _runAcquisitionEbeam(self, future):
                     raise CancelledError()
                 self._acq_state = FINISHED
 
-            for n, das in enumerate(self._acq_data):
-                self._onCompletedData(n, das)
+            self._im_needs_recompute.clear()
+            # Broadcast all the self._LiveData into self._raw and do post-processing
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: I don't think "broadcast" might is the right term. Maybe rephrase as "Adjust the (intermediary) ._live_data to the right shape/format for the final ._raw".

File: src/odemis/acq/stream/_sync.py [hunk 7]
```diff
@@ -1130,8 +1299,11 @@ def _runAcquisitionEbeam(self, future):
...
+            self._raw = [] # reset the ._raw variable
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: That shouldn't be needed, as it's done at the beginning of the function.

File: src/odemis/acq/stream/_sync.py [hunk 8]
```diff
@@ -1741,30 +1913,48 @@ def _runAcquisition(self, future):
...
             leech_np, leech_time_ppx = self._startLeeches(px_time, tot_num, (rep[1], rep[0]))
 
+            self._im_needs_recompute = threading.Event()
+            self._live_update_period = 0.33 # s
+            self._liveImageThread = threading.Thread(target = self._image_thread,
+                                                     args = (weakref.ref(self), self._live_update_period),
+                                                     name = "Live update of the acquisition")
+            self._liveImageThread.daemon = True
+            self._liveImageThread.start()
+
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Can be moved to the `__init__`.

File: src/odemis/acq/stream/_sync.py [hunk 9]
```diff
@@ -1741,30 +1913,48 @@ def _runAcquisition(self, future):
...
             # number of spots scanned so far
             spots_sum = 0
             while spots_sum < tot_num:
+                px_idx = (spots_sum//rep[0], spots_sum%rep[0]) #current pixel index
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Move this line to just before `_current_scan_area`.
BTW, please add spaces around the operators.

File: src/odemis/acq/stream/_sync.py [hunk 10]
```diff
@@ -1741,30 +1913,48 @@ def _runAcquisition(self, future):
...
                 # Acquire the maximum amount of pixels until next leech
-                npixels = min([np for np in leech_np if np is not None] +
-                              [tot_num - spots_sum])  # max, in case of no leech
-                n_y, n_x = leech.get_next_rectangle((rep[1], rep[0]), spots_sum, npixels)
-                npixels = n_x * n_y
+                self._npixels2scan = min([np for np in leech_np if np is not None] +
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: No need (anymore) to store it on self. => You can revert to the original name `npixels`.

File: src/odemis/acq/stream/_sync.py [hunk 11]
```diff
@@ -1741,30 +1913,48 @@ def _runAcquisition(self, future):
...
+                                         [tot_num - spots_sum] + [int(self._live_update_period // px_time + 1)]) # max, in case of no leech
+                #TODO K.K. update unit tests to test for:
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Please do this TODO, or drop this line if it's done now.

File: src/odemis/acq/stream/_sync.py [hunk 12]
```diff
@@ -1741,30 +1913,48 @@ def _runAcquisition(self, future):
...
+                # cornercase where leech is set far after update_interval and therefore a non rectangular block may be scanned
+                # case where leech and update interval run semi in phase/out of phase
+                n_y, n_x = leech.get_next_rectangle((rep[1], rep[0]), spots_sum, self._npixels2scan)
+                self._npixels2scan = n_x * n_y
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Indicate *why* you recompute it afterwards (ie, it might be smaller, to stay a rectangle)

File: src/odemis/acq/stream/_sync.py [hunk 13]
```diff
@@ -1741,30 +1913,48 @@ def _runAcquisition(self, future):
...
+
+                self._current_scan_area = (px_idx[1],
+                                           px_idx[0],
+                                           px_idx[1] + self._npixels2scan,
+                                           px_idx[0])
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: That looks incorrect. -> `px_idx[1] + ny, px_idx[0]+nx`?

File: src/odemis/acq/stream/_sync.py [hunk 14]
```diff
@@ -1806,6 +1994,17 @@ def _runAcquisition(self, future):
                         raise CancelledError()
                     s._dataflow.unsubscribe(self._subscribers[i])
 
+                px_x, px_y = divmod(spots_sum, rep[0])
+                px_idx = (px_x, px_y)
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: That should be merged with px_idx computed at the beginning. (You could just drop these two lines, or use divmod up).

File: src/odemis/acq/stream/_sync.py [hunk 15]
```diff
@@ -1833,17 +2032,19 @@ def _runAcquisition(self, future):
...
                 if self._acq_state == CANCELLED:
                     raise CancelledError()
                 self._acq_state = FINISHED
 
-            # TODO: ideally, we would directly assemble the _acq_data into the final
-            # raw, which would avoid temporarily holding twice all the data.
-
-            for n, das in enumerate(self._acq_data):
-                self._onCompletedData(n, das)
+            self._im_needs_recompute.clear()
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Drop this line, I don't think it's needed.

File: src/odemis/acq/stream/_sync.py [hunk 16]
```diff
@@ -1806,6 +1994,17 @@ def _runAcquisition(self, future):
                         raise CancelledError()
                     s._dataflow.unsubscribe(self._subscribers[i])
 
+                px_x, px_y = divmod(spots_sum, rep[0])
+                px_idx = (px_x, px_y)
+
+                for i, das in enumerate(self._acq_data):
+                    self._assembleLiveData2D(i, das[-1], px_idx, rep, 0)
+
+                # Activate _updateImage thread
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: That's pretty much the exact description of what `shouldUpdateImage()` does. Indicate *why* now and what are the concequences. Maybe something like "Update the live image of each stream".
That said, how does it work for the other streams than the first one? I don't see where is the code...

File: src/odemis/acq/stream/_sync.py [hunk 17]
```diff
@@ -2178,6 +2411,33 @@ class SEMTemporalSpectrumMDStream(SEMCCDMDStream):
...
+            md[MD_POS] = self._live_data[0][pol_idx].metadata[MD_POS]
+            md[MD_DESCRIPTION] = self._streams[n].name.value
+            md[MD_DIMS] = "CTZYX"
+
+            self._live_data[n] = model.DataArray(
+                                 numpy.zeros(shape = (spec_res, temp_res, 1, rep[1], rep[0]), dtype = raw_data.dtype), md)
+
+        self._live_data[n][:, :, 0, px_idx[0], px_idx[1]] = raw_data.reshape(spec_res, temp_res)
+
     def _onCompletedData(self, n, raw_das):
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: All the _onCompletedData() methods and their helper functions need to go away.

File: src/odemis/acq/stream/_sync.py [hunk 18]
```diff
@@ -773,6 +778,143 @@ def _stopLeeches(self):
...
+                       px_idx[1]: px_idx[1] + tile_shape[1]] = True
+        self._live_data[n][pol_idx][
+                           px_idx[0]: px_idx[0] + tile_shape[0],
+                           px_idx[1]: px_idx[1] + tile_shape[1]] = raw_data
+
+    def _assembleFinalData(self, data, pol_idx, pos_polarizations, stream_idx):
+        '''
+        :param data: all acquired data of the streams
+        :param pol_idx: (int) polarisation index related to name as defined in pos_polarizations variable
+        :param pos_polarizations: (list of str) number of possible polarization possibilities
```
[the remark this answers is not in the document]
Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: pos_polarizations we need to not average if there are not multiple polarisations.
pol_idx is very nice if stuff goes wrong for debugging. Since it does not make the function unreadable or messy I propose to leave it in place.

File: src/odemis/acq/stream/_sync.py [hunk 19]
```diff
@@ -1037,17 +1181,17 @@ def _runAcquisitionEbeam(self, future):
...
-                logging.debug("Will acquire the following polarization positions: %s",
-                              pos_polarizations)
+                    logging.debug("Will acquire the following polarization position: %s" % pos_polarizations)
                 # extra time to move pol analyzer for each pos requested (value is very approximate)
                 time_move_pol_once = POL_MOVE_TIME  # s
+                logging.debug("Add %s extra sec to move polarization analyzer for all positions requested."
+                              % time_move_pol_left)
                 time_move_pol_left = time_move_pol_once * len(pos_polarizations)
-                logging.debug("Add %s extra sec to move polarization analyzer for all positions requested.",
-                              time_move_pol_left)
```
[the remark this answers is not in the document]
Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: do you want to lose these logging messages of just change them back as they where?

[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/acq/stream/_sync.py]: Change them back (to the version which is currently in master/HEAD).

File: src/odemis/acq/stream/_sync.py [hunk 20]
```diff
@@ -773,6 +778,143 @@ def _stopLeeches(self):
...
+                       px_idx[1] * tile_shape[1]:(px_idx[1] + 1) * tile_shape[1]] = True
+        self._live_data[n][pol_idx][
+                       px_idx[0] * tile_shape[0]:(px_idx[0] + 1) * tile_shape[0],
+                       px_idx[1] * tile_shape[1]:(px_idx[1] + 1) * tile_shape[1]] = raw_data
+
+    def _assembleLiveData2D(self, n, raw_data, px_idx, rep, pol_idx):
+        """
+        :param n: (int) number of the current stream
+        :param raw_data: acquired data of stream
+        :param px_idx: (tuple of int) pixel index: y, x
```
[the remark this answers is not in the document]
Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: add to docstring that pxidx is first pixel of last aquisition of CL since CL works different

File: src/odemis/acq/stream/_sync.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: ok remove

File: src/odemis/acq/stream/_sync.py [hunk 21]
```diff
@@ -1062,18 +1206,32 @@ def _runAcquisitionEbeam(self, future):
...
             n = 0  # number of images acquired so far
 
-            for pol_pos in pos_polarizations:
+            self._im_needs_recompute = threading.Event()
+            self._live_update_period = 0.33  # s
+            self._liveImageThread = threading.Thread(target = self._image_thread,
+                                                     args = (weakref.ref(self), self._live_update_period),
+                                                     name = "Live update of the acquisition")
+            self._liveImageThread.daemon = True
+            self._liveImageThread.start()
```
[the remark this answers is not in the document]
Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: fix as in todo

File: src/odemis/acq/stream/_sync.py (hunk 11 above)
Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: remove K.K.

File: src/odemis/acq/stream/_sync.py [hunk 22]
```diff
@@ -834,62 +859,48 @@ def _assembleLiveData2D(self, n, raw_data, px_idx, rep, pol_idx):
...
+        Create RGB projection of over of live SEM data, also adds a blue background of non-scanned pixels and orange
+        pixels for the pixels which are currently being scanned.
         """
         # Here we need to pass a *copy* of mask, because it might change *while* numpy is doing
         # the indexing. It seems that numpy first allocates the size of data_acq before
         # actually copying the values, so if there are eventually more values, it can overflow
         # the array index.
-        data_acq = data[self._acq_mask.copy()]
-        logging.debug("Got %s data for histogram", data_acq.shape)
+        acq_mask = self._acq_mask.copy() #because of threading issues this variable needs to be copied
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Please drop this comment, and just update the explanation just above.

File: src/odemis/acq/stream/_sync.py [hunk 23]
```diff
@@ -899,21 +910,35 @@ def _projectXY2RGB(self, data, tint=(255, 255, 255)):
...
-        self.streams[0].raw = [self._live_data[0][-1]]  # For GetBoundingBox()
-        no_acq_mask = numpy.invert(self._acq_mask)
-        scan_area = self._current_scan_area
-
-        rgbim = self._projectXY2RGB(self.streams[0].raw[0])
-        rgbim = self._progressPixels(rgbim, scan_area, no_acq_mask)
+        try:
+            raw_data = self._live_data[0][-1]
+        except IndexError:  # Can happen if the acquisition has just finished
+            if self._acq_done.isSet():
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Please use the "new" name which is `is_set()`.

File: src/odemis/acq/stream/_sync.py [hunk 24]
```diff
@@ -899,21 +910,35 @@ def _projectXY2RGB(self, data, tint=(255, 255, 255)):
...
+                logging.debug("Not updating live image, as acquisition is over")
+                return
+            else:
+                raise
 
-        rgbim.flags.writeable = False
+        self.streams[0].raw = [raw_data]  # For GetBoundingBox()
+        rgbim = self._projectXY2RGB(raw_data)
+        # Don't update if the acquisition is already over
+        if self._acq_done.isSet():
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: -> `is_set()`

File: src/odemis/acq/stream/_sync.py [hunk 25]
```diff
@@ -1636,6 +1659,12 @@ def _runAcquisitionScanStage(self, future):
 
             n = 0  # number of points acquired so far
             for px_idx in numpy.ndindex(*rep[::-1]):  # last dim (X) iterates first
+                #TODO K.K. what is both CL and SEM-scanstage are used simoultounously what is displayed? what is called?
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: I'm not exactly sure I understand the question, but essentially, the same place is scanned as in the equivalent code in `runAcquisitionEbeam()`. So that code looks fine.

File: src/odemis/acq/stream/_sync.py [hunk 26]
```diff
@@ -1790,8 +1826,12 @@ def _runAcquisitionScanStage(self, future):
...
                 self._acq_state = FINISHED
 
-            for n, das in enumerate(self._acq_data):
-                self._onCompletedData(n, das)
+            # Process all the (intermediary) ._live_data to the right shape/format for the final ._raw
+            for stream_idx, das in enumerate(self._live_data):
+                self._assembleFinalData(stream_idx, das)
+
+            # for n, das in enumerate(self._acq_data):
+            #     self._onCompletedData(n, das)
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Drop these lines.

File: src/odemis/acq/stream/_sync.py [hunk 27]
```diff
@@ -1923,38 +1967,29 @@ def _runAcquisition(self, future):
...
+                npixels2scan = min([np for np in leech_np if np is not None] +
                                          [tot_num - spots_sum] + [int(self._live_update_period // px_time + 1)]) # max, in case of no leech
-                #TODO K.K. update unit tests to test for:
+                #TODO update unit tests to test for:
                 # cornercase where leech is set far after update_interval and therefore a non rectangular block may be scanned
                 # case where leech and update interval run semi in phase/out of phase
-                n_y, n_x = leech.get_next_rectangle((rep[1], rep[0]), spots_sum, self._npixels2scan)
-                self._npixels2scan = n_x * n_y
+                n_y, n_x = leech.get_next_rectangle((rep[1], rep[0]), spots_sum, npixels2scan) #method to only scan
+                # rectangular blocks of pixels which start at the beginning of a row
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Either put the long comment in a separate line *before* the function call, or make the comment fit on the line. Maybe you can just write `# also ensure it's a rectangle`

File: src/odemis/acq/stream/_sync.py [hunk 28]
```diff
@@ -2072,10 +2102,35 @@ def _runAcquisition(self, future):
...
             self._dc_estimator = None
             self._current_future = None
             self._acq_done.set()
 
+    def _updateImage(self):
+        """
+        Function called by image update thread which handles updating the overlay of the SEM live update image
+        """
+        # Display only the SEM image taken in the last polarization and convert to RGB + display
+        for stream in range(2):
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: I'd suggest to write it like `for i, stream in enumerate(self._stream)`.
The, you can replace the `stream` -> `i`, and `self.streams[stream]` -> `stream`.

File: src/odemis/gui/cont/acquisition.py [hunk 29]
```diff
@@ -815,9 +817,11 @@ def on_acquisition_done(self, future):
             logging.error("Acquisition failed (after %d streams): %s",
                           len(data), exp)
 
-        # Remove live update
+        #  REMOVE the the overlay (live_update) in the SEM window which displays the SEM measurements of the current
+        #  acquisition if a  stream is added and acquisition is started.
+        # TODO K.K. fix that also the CL stream is removed
```
Comment (pieleric) [inline src/odemis/gui/cont/acquisition.py]: No, I think the CL stream should stay, as it's a setting stream. That said, you might want to make sure the last data displayed corresponds to the complete acquisition (without any orange/blue indicators).... but that not a big deal if it's not done and just left as a TODO.

File: src/odemis/acq/stitching/_registrar.py [hunk 30]
```diff
@@ -206,8 +206,11 @@ def getPositions(self):
         tile_positions = []
         dep_tile_positions = []
 
+        # TODO K.K. remove this or implement differently
+        self.shift_wilco = [] #Made for Wilco, currenlty a debug value
         for ti in self.acqOrder:
             shift = self.registered_positions[ti[0]][ti[1]]
+            self.shift_wilco.append(shift)
```
[the remark this answers is not in the document]
Comment (Kleijwegt) [reply src/odemis/acq/stitching/_registrar.py]: shit these are changes for Sonic - mixed up commits

File: src/odemis/acq/stream/_sync.py [hunk 31]
```diff
@@ -789,6 +802,10 @@ def _assembleLiveDataTiles(self, n, raw_data, px_idx, rep, pol_idx):
         :param rep: (tuple of int) repetition frame/ size of entire frame
         :param pol_idx: (int) polarisation index related to name as defined in pos_polarizations variable
         uses fuzzing
+        One full SEM scan is stored for each polarisation, at the end assembleFinalData() will take care of
+        integrating (aka averaging) them into a single one. To do this "live" would be diffucult and prone to error's since
+        an weighted average over the total number of polarisations in combination with the previous done
+        polarisations need te be made.
```
[the remark this answers is not in the document]
Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: It feels a bit odd to start the function with a comment since this is a general comment. Where do you suggest to put this comment?

[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/acq/stream/_sync.py]: I think that it's fine to be at the start of the function. BTW, I'd suggest adjust the punctuation to read more easily:
" One full SEM scan is stored for each polarisation. At the end, assembleFinalData() will take care of..."

File: src/odemis/acq/stream/_base.py [hunk 32]
```diff
@@ -848,43 +848,40 @@ def _shouldUpdateImage(self):
...
             name = stream.name.value
             im_needs_recompute = stream._im_needs_recompute
             # Only hold a weakref to allow the stream to be garbage collected
             # On GC, trigger im_needs_recompute so that the thread can end too
             wstream = weakref.ref(stream, lambda o: im_needs_recompute.set())
-
-            tnext = 0
             while True:
                 del stream
+                #TODO K.K. Eric does this go alright with waiting/pointers etc?
```
Comment (pieleric) [inline src/odemis/acq/stream/_base.py]: That looks fine. Especially, assuming it was fine, the changes shouldn't bring more issues.

File: src/odemis/acq/stream/_sync.py [hunk 33]
```diff
@@ -1183,7 +1375,7 @@ def _integrateImages(self, n):
             orig_dtype = data2process[0].dtype
 
             if det_type == model.MD_DT_NORMAL:  # SEM
-                logging.debug("Averaging %s images", n)
+                # logging.debug("Average %s images", n)
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Please leave these log as they were (l 1378 and 1387).

File: src/odemis/acq/stream/_sync.py [hunk 34]
```diff
@@ -1600,6 +1802,13 @@ def _runAcquisitionScanStage(self, future):
                             leech_np[li] = np
                             if self._acq_state == CANCELLED:
                                 raise CancelledError()
+                    #TODO K.K. just added code here for scan stage, check if placed right
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: That looks alright. => remove this comment.

**Comment by Kleijwegt** [comment]
> Just added a few extra comments, mostly concerning the "TODO k.k." that you had left. Please have a look at them, and the few other comments left, and let's try to get this merged soon.

hjfgjyjy

File: src/odemis/acq/stream/_sync.py (hunk 34 above)
Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: oeps that was actually an comment for my-future-selfto review my-past-self

**Comment by Kleijwegt** [comment]
@pieleric I have implemented a monochrometer fix and a fix for the ugly orange line/dot in the CL update on the last run (this was easy).
Do you like this implementation?
Do you know how to propely wait on the thread (which updates the live view) to finish so instead of using time.sleep()?

File: src/odemis/acq/stream/_sync.py [hunk 35]
```diff
@@ -1833,17 +2068,31 @@ def _runAcquisition(self, future):
...
+                if isinstance(self._streams[stream_idx], MonochromatorSettingsStream):
+                    # The Monochromator stream uses a chronogram view, which is not compatible with the RGB spatial image
+                    continue
+                elif stream_idx >= 1:
+                    # Only update the CL stream
+                    self._streams[1].raw = [self._live_data[1][-1]]  # For GetBoundingBox()
+                    rgbim = self._projectXY2RGB(self._live_data[1][-1])
+                    self._streams[1].image.value = rgbim
+                    #TODO K.K. what is better? Above or below?
+                    # self._streams[stream_idx]._onNewData(self._streams[stream_idx]._dataflow, self._raw[-1])
```
Comment (Kleijwegt) [inline src/odemis/acq/stream/_sync.py]: @pieleric I am not sure if my implementation is actually that much better, which one do you prefer?

Comment (Kleijwegt) [reply src/odemis/acq/stream/_sync.py]: after this I will remove all the _Oncompletedata, remove all the unnecacerry changes etc. and then this should be ready to be merged.

Comment (pieleric) [reply src/odemis/acq/stream/_sync.py]: I think that if you end-up still needing this "isinstance()" (because you cannot call `updateImage()` directly, then just go for the shorter version, with `onNewData()`.
BTW, I'd suggest to put this block concerning the image update into a try/except block, and only log the exceptions, because we don't want any issue here to cause a complete failure of the acquisition.

**Comment by pieleric** [comment]
Once you're done adjusting this small bit, I think the code should be ready. You'll still need to rebase with upstream (to fix the conflict), remove the onCompleteData() functions, put back the testing code, and remove the extra changes related to the stitching.
Finally, please squash you changes, and put a nice commit text explanation.

File: plugins/cliccd.py [hunk 36]
```diff
@@ -60,16 +61,22 @@ def _preprocessData(self, n, data, i):
...
-            return r
+            return super(SEMCLCCDStream, self)._assembleFinalData(n, data)
 
         # Same as sem data, but without computing the data position from the
         # CCD metadata
         md = self._ccd_md.copy()
-        sem_data = self._raw[0]  # _onCompletedData() should be called in order
+        sem_data = self._raw[0]  # _assembleFinalData() should be called in order
         md[model.MD_POS] = sem_data.metadata[model.MD_POS]
         md[model.MD_DESCRIPTION] = self._streams[n].name.value
```
Comment (pieleric) [inline plugins/cliccd.py]: You can drop this line, as you've put it in `_assembleLiveData()`.
I think that could also move the MD_POS and removal of the MD_AR* to `_assembleLiveData()`.

File: plugins/cliccd.py [hunk 37]
```diff
@@ -90,7 +97,7 @@ def _onCompletedData(self, n, raw_das):
             logging.warning("Metadata missing from the SEM data")
```
[the remark this answers is not in the document]
Comment (pieleric) [reply plugins/cliccd.py:95]: It looks alright. It essentially does the same as `get_center_pxs()`.
BTW, I guess in theory, we should move more of this code to the "assembleLiveData()", as every new pixel (ie, CCD image) could immediately be stored in the final array. Anyway, this is good enough...

**Comment by Kleijwegt** [comment]
@pieleric Ready to be merged

**Comment by Kleijwegt** [comment]
@pieleric ready to be merged
